From 684130a65a6416a33525602e249099cb0d65199b Mon Sep 17 00:00:00 2001
Message-Id: <684130a65a6416a33525602e249099cb0d65199b.1417780878.git.chang-joon.lee@intel.com>
In-Reply-To: <cb098d33cdac3271103834fbce08218248d7bddb.1417780878.git.chang-joon.lee@intel.com>
References: <cb098d33cdac3271103834fbce08218248d7bddb.1417780878.git.chang-joon.lee@intel.com>
From: Gajanan Bhat <gajanan.bhat@intel.com>
Date: Wed, 30 Jul 2014 17:26:11 +0530
Subject: [PATCH 11/34] FOR_UPSTREAM [VPG]: drm/i915: Use crtc as parameter
 for zorder

Use crtc ID as parameter to control which pipe zorder has
to change.

Issue: GMINL-1142
Change-Id: I1395ec9d83fafc8bab769f949901ee03d3a23fbd
Signed-off-by: Gajanan Bhat <gajanan.bhat@intel.com>
Signed-off-by: Vidya Srinivas <vidya.srinivas@intel.com>
Signed-off-by: Sangani Suryanarayana Raju <suryanarayana.r.sangani@intel.com>
---
 drivers/gpu/drm/i915/intel_display.c |    1 +
 drivers/gpu/drm/i915/intel_sprite.c  |   15 ++++++++++++---
 include/uapi/drm/i915_drm.h          |    1 +
 3 files changed, 14 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/drm/i915/intel_display.c b/drivers/gpu/drm/i915/intel_display.c
index ac74b20..d4982f9 100644
--- a/drivers/gpu/drm/i915/intel_display.c
+++ b/drivers/gpu/drm/i915/intel_display.c
@@ -10718,6 +10718,7 @@ int intel_set_disp_calc_flip(struct drm_mode_set_display *disp,
 			return -ENOMEM;
 		}
 		zorder->order = disp->zorder;
+		zorder->obj_id = disp->crtc_id;
 		tmp_ret = i915_set_plane_zorder(dev, (void *)zorder, NULL);
 		if (tmp_ret) {
 			DRM_ERROR("i915_set_plane_zorder failed\n");
diff --git a/drivers/gpu/drm/i915/intel_sprite.c b/drivers/gpu/drm/i915/intel_sprite.c
index 650be2a..0a4ba5e 100644
--- a/drivers/gpu/drm/i915/intel_sprite.c
+++ b/drivers/gpu/drm/i915/intel_sprite.c
@@ -315,12 +315,21 @@ int i915_set_plane_zorder(struct drm_device *dev, void *data,
 	struct drm_i915_private *dev_priv = dev->dev_private;
 	u32 val = 0;
 	struct drm_i915_set_plane_zorder *zorder = data;
+	struct drm_mode_object *obj;
+	struct intel_crtc *intel_crtc;
+	int pipe;
 	u32 order = zorder->order;
 	int s1_zorder, s1_bottom, s2_zorder, s2_bottom;
-	int pipe = (order >> 31) & 0x1;
 
-	struct intel_crtc *intel_crtc =
-			to_intel_crtc(dev_priv->plane_to_crtc_mapping[pipe]);
+	obj = drm_mode_object_find(dev, zorder->obj_id, DRM_MODE_OBJECT_CRTC);
+	if (!obj) {
+		DRM_ERROR("Unknown CRTC ID: %lu\n",
+				(unsigned long)zorder->obj_id);
+		return -EINVAL;
+	}
+
+	intel_crtc = to_intel_crtc(obj_to_crtc(obj));
+	pipe = intel_crtc->pipe;
 
 	s1_zorder = (order >> 3) & 0x1;
 	s1_bottom = (order >> 2) & 0x1;
diff --git a/include/uapi/drm/i915_drm.h b/include/uapi/drm/i915_drm.h
index 430c664..b0bc4f8 100644
--- a/include/uapi/drm/i915_drm.h
+++ b/include/uapi/drm/i915_drm.h
@@ -1161,6 +1161,7 @@ struct drm_i915_reg_read {
 
 
 struct drm_i915_set_plane_zorder {
+	__u32 obj_id;
 	__u32 order;
 };
 
-- 
1.7.9.5

